lib/api/apiUtils/authorization/validation.js

```diff
@@ -0,0 +1,217 @@
+import { isBucketAuthorized, isObjAuthorized } from './aclChecks';
+import bucketShield from '../bucket/bucketShield';
+
+/** metadataGetBucket - retrieves bucket from metadata, returning error if
+ * bucket is shielded
+ * @param {string} requestType - type of request
+ * @param {string} bucketName - name of bucket
+ * @param {RequestLogger} log - request logger
+ * @param {function} cb - callback
+ * @return {undefined} - and call callback with err, bucket md
+ */
+export function metadataGetBucket(requestType, bucketName, log, cb) {
+    return metadata.getBucket(bucketName, log, (err, bucket) => {
+        if (err) {
+            log.debug('metadata getbucket failed', { error: err });
+            return cb(err);
+        }
+        if (bucketShield(bucket, requestType)) {
+            log.debug('bucket is shielded from request', {
+                requestType,
+                method: 'metadataGetBucketAndObject',
+            });
+            // still return bucket for cors headers
+            return cb(errors.NoSuchBucket, bucket);
+        }
+        log.trace('found bucket in metadata');
+        return cb(null, bucket);
+    });
+}
+
+/** metadataValidateBucket - retrieve bucket from metadata and check if user
+ * is authorized to access it
+ * @param {object} params - function parameters
+ * @param {AuthInfo} params.authInfo - AuthInfo class instance, requester's info
+ * @param {string} params.bucketName - name of bucket
+ * @param {string} params.requestType - type of request
+ * @param {RequestLogger} log - request logger
+ * @param {function} callback - callback
+ * @return {undefined} - and call callback with params err, bucket md
+ */
+export function metadataValidateBucket(params, log, callback) {
+    const { authInfo, bucketName, requestType } = params;
+    const canonicalID = authInfo.getCanonicalID();
+    return metadataGetBucket(requestType, bucketName, log, (err, bucket) => {
+        if (err) {
+            // still return bucket for cors headers
+            return callback(err, bucket);
+        }
+        if (!isBucketAuthorized(bucket, requestType, canonicalID)) {
+            log.debug('access denied for user on bucket', { requestType });
+            return callback(errors.AccessDenied, bucket);
+        }
+        return callback(null, bucket);
+    });
+}
+
+/** getNullVersion - return metadata of null version if it exists
+ * @param {object} objMD - metadata of master version
+ * @param {string} bucketName - name of bucket
+ * @param {string} objectKey - name of object key
+ * @param {RequestLogger} log - request logger
+ * @param {function} cb - callback
+ * @return {undefined} - and call callback with params err, objMD of null ver
+ */
+export function getNullVersion(objMD, bucketName, objectKey, log, cb) {
+    const options = {};
+    if (objMD.isNull || !objMD.versionId) {
+        // null version is current version
+        log.debug('found null version'); // don't want to log internal id?
```

Comment:
the log doesn't currently log the decrypted version id so I will keep it
but in general we can consider whether it would be useful to log the version id, and if so, only when it is encrypted or the 'internal' (decrypted) vid as well? it's something we can save for later; I'll remove the comment